upload.py

```diff
@@ -33,55 +35,80 @@ def hrsize(size):
             return '%.0f%sB' % (size, suffix)
     return '%.0f%sB' % (size, 'Y')
 
+
+class HashingFile(file):
+    def __init__(self, file_path):
+        super(HashingFile, self).__init__(file_path, "w+b")
+        self.sha1 = hashlib.sha1()
+
+    def write(self, data):
+        self.sha1.update(data)
+        return file.write(self, data)
+
+    def get_hash(self):
+        return self.sha1.hexdigest()
+
+
+def getHashingFieldStorage(upload_dir):
+    class HashingFieldStorage(cgi.FieldStorage):
```

Comment:
Think of this as a factory pattern. Since the upload_dir is required to create a HashingFieldStorage instance, I don't want to expose the raw class. This encapsulates that complexity.